modules/internal/ChapelDistribution.chpl

```diff
@@ -90,6 +90,10 @@ module ChapelDistribution {
       return (count==0);
     }
 
+    //
+    // TODO: Are there opportunities to optimize out the locking here,
+    // as on the add_arr() case?
```

Comment:
I'll both answer the question and request you add the answer to the comment.
We could optimize for example the construction of a domain rank change because that operation creates a new distribution. It's unclear how important this optimization is, though, because domain rank change is much less common than array slicing. However, adding the optimization would amount to adding a default argument e.g. `lockingDist=true` to `newRectangularDom`; then passing that to `add_dom`; and finally updating domain rank change `this` method to pass `lockingDist=false` to `newRectangularDom`.

AFAIK this domain rank change is the only case we could improve by doing this.